File: ruby/day19_part1_2020.rb

rules, messages = File.read("input.txt").split("\n\n")

rules = rules.split("\n").map { |rule| rule.split(": ") }
messages = messages.split("\n")

def build_regex(rules, rule_id, rules_hash)
  return rules_hash[rule_id] if rules_hash[rule_id]

  rule = rules.find { |r| r[0] == rule_id }[1]

  if rule.include?("\"")
    rules_hash[rule_id] = rule.delete("\"")
  else
    sub_rules = rule.split(" | ").map do |sub_rule|
      sub_rule.split(" ").map do |id|
        build_regex(rules, id, rules_hash)
      end.join
    end

    rules_hash[rule_id] = "(#{sub_rules.join("|")})"
  end
end

rules_hash = {}
build_regex(rules, "0", rules_hash)

regex = Regexp.new("^#{rules_hash["0"]}$")

puts messages.count { |message| regex.match?(message) }

File: lua/day19_part1_2020.lua
local function parse_rule(line)
    local id, rule = line:match("(%d+): (.+)")
    id = tonumber(id)
    if rule:find('"') then
        return id, rule:sub(2, -2) -- Direct character match, removing quotes
    else
        local subrules = {}
        for part in rule:gmatch("[^|]+") do
            local subrule = {}
            for num in part:gmatch("%d+") do
                table.insert(subrule, tonumber(num))
            end
            table.insert(subrules, subrule)
        end
        return id, subrules
    end
end

local function read_input(filename)
    local rules = {}
    local messages = {}
    local reading_rules = true
    for line in io.lines(filename) do
        if line == "" then
            reading_rules = false
        elseif reading_rules then
            local id, rule = parse_rule(line)
            rules[id] = rule
        else
            table.insert(messages, line)
        end
    end
    return rules, messages
end

local function match_rule(rules, rule_id, message, start)
    local rule = rules[rule_id]
    if type(rule) == "string" then
        return message:sub(start, start) == rule and start + 1
    else
        for _, subrule in ipairs(rule) do
            local pos = start
            local matches = true
            for _, sub in ipairs(subrule) do
                pos = match_rule(rules, sub, message, pos)
                if not pos then
                    matches = false
                    break
                end
            end
            if matches then
                return pos
            end
        end
        return nil
    end
end

local function count_valid_messages(rules, messages)
    local count = 0
    for _, message in ipairs(messages) do
        local match_end = match_rule(rules, 0, message, 1)
        if match_end and match_end - 1 == #message then
            count = count + 1
        end
    end
    return count
end

local rules, messages = read_input("input.txt")
local valid_count = count_valid_messages(rules, messages)
print("Number of valid messages:", valid_count)